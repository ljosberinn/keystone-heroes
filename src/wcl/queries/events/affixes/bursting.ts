import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const BURSTING = {
  damage: 243_237,
  debuff: 240_443,
};

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=4&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22damage%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20243237%20and%20rawDamage%20%3E%200
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 5987472, endTime: 7736379, limit: 2000, filterExpression: "type = \"damage\" and target.type = \"player\" and ability.id = 243237 and rawDamage > 0") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "damage" and target.type = "player" and ability.id = ${BURSTING.damage} and rawDamage > 0`,
  `type in ("applydebuff", "applydebuffstack") and ability.id = ${BURSTING.debuff}`,
];

const isBurstingDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: BURSTING.damage,
});

const isBurstingApplyDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: BURSTING.debuff,
});

const isBurstingApplyDebuffStackEvent =
  createIsSpecificEvent<ApplyDebuffStackEvent>({
    type: "applydebuffstack",
    abilityGameID: BURSTING.debuff,
  });

export const getBurstingEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | ApplyDebuffEvent | ApplyDebuffStackEvent)[] => {
  if (!affixSet.has(Affixes.Bursting)) {
    return [];
  }

  return [
    ...allEvents.filter(isBurstingDamageEvent),
    ...allEvents.filter(isBurstingApplyDebuffEvent),
    ...allEvents.filter(isBurstingApplyDebuffStackEvent),
  ];
};
