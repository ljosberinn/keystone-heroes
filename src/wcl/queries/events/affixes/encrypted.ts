import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const encryptedDebuffIDs = {
  // Urh
  cdrRegen: 368_239,
  // Wo
  msStealth: 368_241,
  // Vy
  hasteDmg: 368_240,
};

export const encryptedDebuffs = [
  {
    name: "Decrypted Urh Cypher",
    id: encryptedDebuffIDs.cdrRegen,
    icon: "inv_progenitor_anima_blue",
    type: "cdr & regen",
  },
  {
    name: "Decrypted Wo Cypher",
    id: encryptedDebuffIDs.msStealth,
    icon: "inv_progenitor_anima_purple",
    type: "ms & stealth",
  },
  {
    name: "Decrypted Vy Cypher",
    id: encryptedDebuffIDs.hasteDmg,
    icon: "inv_progenitor_anima_green",
    type: "haste & dmg",
  },
];

export const encryptedMinibosses = [
  {
    name: "Urh Dismantler",
    id: 184_911,
  },
  {
    name: "Wo Drifter",
    id: 184_910,
  },
  {
    name: "Vy Interceptor",
    id: 184_908,
  },
];

export const encryptedAbilityIDs = {
  // urh relic
  energyBarrage: 368_077,
  // vy
  shoot: 366_406,
  fusionBeam: 366_409,
  vyCypher: 368_495, // on hit proc
  // Wo
  burst: 366_566,
  stealth: 368_162,
  // Urh
  deconstruct: 366_297,
};

export const encryptedAbilities = [
  {
    name: "Energy Barrage",
    id: encryptedAbilityIDs.energyBarrage,
    icon: "spell_progenitor_missile",
  },
  {
    name: "Shoot",
    id: encryptedAbilityIDs.shoot,
    icon: "spell_progenitor_missile",
  },
  {
    name: "Fusion Beam",
    id: encryptedAbilityIDs.fusionBeam,
    icon: "spell_progenitor_beam",
  },
  {
    name: "Burst",
    id: encryptedAbilityIDs.burst,
    icon: "spell_progenitor_areadenial",
  },
  {
    name: "Deconstruct",
    id: encryptedAbilityIDs.deconstruct,
    icon: "ability_warrior_shieldbreak",
  },
  {
    name: "Wo Cloaking Field",
    id: encryptedAbilityIDs.stealth,
    icon: "ability_vanish",
  },
  {
    name: "Decrypted Vy Cypher",
    id: encryptedAbilityIDs.vyCypher,
    icon: "spell_broker_buff",
  },
];

export const encryptedRelics = [
  {
    name: "Wo Relic",
    id: 185_683,
  },
  {
    name: "Vy Relic",
    id: 185_680,
  },
  {
    name: "Urh Relic",
    id: 185_685,
  },
];

export const filterExpression = [
  `type = "applydebuff" and ability.id in (${Object.values(
    encryptedDebuffIDs
  ).join(", ")})`,
  `type = "damage" and ability.id in (${Object.values(encryptedAbilityIDs).join(
    ", "
  )})`,
];

const isEncryptedDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: Object.values(encryptedDebuffIDs),
});

const isEncryptedDamageOrDamgeTakenEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: Object.values(encryptedAbilityIDs),
});

export const getEncryptedEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | ApplyDebuffEvent)[] => {
  if (!affixSet.has(Affixes.Encrypted)) {
    return [];
  }

  return allEvents.reduce<(DamageEvent | ApplyDebuffEvent)[]>((acc, event) => {
    if (
      isEncryptedDamageOrDamgeTakenEvent(event) ||
      isEncryptedDebuffEvent(event)
    ) {
      acc.push(event);
    }

    return acc;
  }, []);
};
