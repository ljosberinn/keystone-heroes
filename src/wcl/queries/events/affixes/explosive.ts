import { Affixes } from "@prisma/client";

import type { AllTrackedEventTypes, DamageEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const EXPLOSIVE = {
  unit: 120_651,
  ability: 240_446,
} as const;

export const EXPLOSIVE_HEALTH = 224 as const;

/**
 * @see https://www.warcraftlogs.com/reports/YrA4zyZgQJvHbn32#fight=1&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24target.id%20%3D120651%20AND%20type%20%3D%20%22damage%22%20AND%20overkill%20%3E%200%5E2%24Off%24%23909049%24expression%24target.type%20%3D%20%22player%22%20and%20rawDamage%20%3E%200%20and%20ability.id%20%3D%20240446
 * ```gql
 * {
 *   reportData {
 *     report(code: "YrA4zyZgQJvHbn32") {
 *       fights(fightIDs: [1]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 87034, endTime: 1527765, limit: 2000, filterExpression: "(target.type = \"player\" and rawDamage > 0 and ability.name = \"Explosion\") OR (target.name = \"Explosives\" AND type = \"damage\" AND overkill > 0)") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `target.type = "player" and rawDamage > 0 and ability.id = ${EXPLOSIVE.ability}`,
  `target.id = ${EXPLOSIVE.unit} and overkill > 0`,
];

const isExplosiveDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: EXPLOSIVE.ability,
});

const createIsExplosiveDeathEvent =
  (targetID: number) =>
  (event: AllTrackedEventTypes): event is DamageEvent => {
    return (
      event.type === "damage" &&
      "targetInstance" in event &&
      event.sourceID !== -1 &&
      event.targetID === targetID &&
      typeof event.overkill !== "undefined" &&
      event.overkill > 0
    );
  };

/**
 * explosives detection is solely based on the hope that it's the unit that
 * occured the most over the course of a logging session, and thus has the highest
 * `targetInstance` property
 */
export const findExplosiveTargetID = (
  allEvents: AllTrackedEventTypes[]
): number | null => {
  const dataset = allEvents.reduce<{
    targetID: null | number;
    targetInstance: null | number;
  }>(
    (acc, event) => {
      if (
        // skip non-DamageEvents
        event.type !== "damage" ||
        // ignore event if target is single instance
        !("targetInstance" in event) ||
        // skip empty instances, just TS here
        event.targetInstance === undefined ||
        // check whether the stored instance is lower than this events
        (acc.targetInstance && event.targetInstance < acc.targetInstance)
      ) {
        return acc;
      }

      return {
        targetID: event.targetID,
        targetInstance: event.targetInstance,
      };
    },
    {
      targetID: null,
      targetInstance: null,
    }
  );

  return dataset.targetID;
};

export const getExplosiveEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): { explosiveTargetID: null | number; events: DamageEvent[] } => {
  if (!affixSet.has(Affixes.Explosive)) {
    return { explosiveTargetID: null, events: [] };
  }

  const explosiveTargetID = findExplosiveTargetID(allEvents);

  if (!explosiveTargetID) {
    return { explosiveTargetID: null, events: [] };
  }

  const explosiveKills = allEvents.filter(
    createIsExplosiveDeathEvent(explosiveTargetID)
  );
  const explosiveDamage = allEvents.filter(isExplosiveDamageEvent);

  return {
    explosiveTargetID,
    events: [...explosiveDamage, ...explosiveKills],
  };
};
