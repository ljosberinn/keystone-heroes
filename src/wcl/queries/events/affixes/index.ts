import { Affixes } from "@prisma/client";

import type { EventParams } from "..";
import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
  RemoveBuffEvent,
  RemoveDebuffEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
  | "Infernal"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
  [Affixes.Encrypted]: [],
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): {
  explosiveTargetID: number | null;
  events: (
    | ApplyBuffEvent
    | ApplyBuffStackEvent
    | DamageEvent
    | HealEvent
    | InterruptEvent
    | ApplyDebuffStackEvent
    | BeginCastEvent
    | CastEvent
    | ApplyDebuffEvent
    | RemoveBuffEvent
    | RemoveDebuffEvent
  )[];
} => {
  const affixSet = new Set(affixes);

  const { explosiveTargetID, events: explosiveEvents } = getExplosiveEvents(
    allEvents,
    affixSet
  );

  return {
    explosiveTargetID,
    events: [
      // seasonal
      ...getTormentedEvents(allEvents, affixSet),
      // common affixes
      ...getStormingEvents(allEvents, affixSet),
      ...getSpitefulEvents(allEvents, affixSet),
      ...getSanguineEvents(allEvents, affixSet),
      ...getVolcanicEvents(allEvents, affixSet),
      ...getQuakingEvents(allEvents, affixSet),
      ...getBolsteringEvents(allEvents, affixSet),
      ...getBurstingEvents(allEvents, affixSet),
      ...explosiveEvents,
      ...getGrievousEvents(allEvents, affixSet),
      ...getNecroticEvents(allEvents, affixSet),
    ],
  };
};
