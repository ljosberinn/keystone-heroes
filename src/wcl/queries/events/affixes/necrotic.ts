import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const NECROTIC = 209_858;

/**
 * @see https://www.warcraftlogs.com/reports/DmVgxdj3WTYBApQt/#fight=1&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24ability.id%20%3D%20209858%20and%20type%20%3D%20%22applydebuffstack%22%20and%20stack%20%3E%2010%5E2%24Off%24%23909049%24expression%24type%20%3D%20%22damage%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20209858%20and%20rawDamage%20%3E%200
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "DmVgxdj3WTYBApQt") {
 *       fights(fightIDs: [1]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 0, endTime: 1414930, filterExpression: "(target.type = \"player\" and ability.id = 209858) and ((type = \"applydebuffstack\" and stack > 10) or (type = \"damage\" and rawDamage > 0))") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */

export const filterExpression = [
  `type in ("applydebuff", "damage", "applydebuffstack") and ability.id = ${NECROTIC}`,
];

const isNecroticDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NECROTIC,
});

const isNecroticDebuffStackEvent = createIsSpecificEvent<ApplyDebuffStackEvent>(
  {
    type: "applydebuffstack",
    abilityGameID: NECROTIC,
  }
);

const isNecroticDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: NECROTIC,
});

export const getNecroticEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | ApplyDebuffStackEvent | ApplyDebuffEvent)[] => {
  if (!affixSet.has(Affixes.Necrotic)) {
    return [];
  }

  return [
    ...allEvents.filter(isNecroticDebuffStackEvent),
    ...allEvents.filter(isNecroticDamageEvent),
    ...allEvents.filter(isNecroticDebuffEvent),
  ];
};
