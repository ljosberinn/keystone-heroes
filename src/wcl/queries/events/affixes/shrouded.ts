import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  CastEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const shroudedNathrezimInfiltratorID = 189_878;
export const shroudedZulgamuxID = 190_128;

export const defaultBrokerRestorationID = 374_295;
export const zulgamuxBrokerRestorationID = 374_300;

export const shroudedAbilities = [
  {
    name: "Shadow Claws",
    id: 373_509,
    icon: "ability_creature_disease_05",
  },
  {
    name: "Shadow Eruption",
    id: 373_513,
    icon: "spell_shadow_painandsuffering",
  },
  {
    name: "Vampiric Claws",
    id: 373_364,
    icon: "ability_druid_disembowel",
  },
  {
    name: "Carrion Swarm",
    id: 373_429,
    icon: "spell_shadow_carrionswarm",
  },
  {
    name: "Blood Siphon", // cast
    id: 373_729,
    icon: "ability_ironmaidens_bloodritual",
  },
  {
    name: "Blood Siphon", // damage source
    id: 373_744,
    icon: "ability_ironmaidens_bloodritual",
  },
  {
    name: "Restoration",
    id: 374_300,
    icon: "spell_broker_buff",
  },
  {
    name: "Restoration",
    id: 374_295,
    icon: "spell_broker_buff",
  },
];

const isShroudedEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: shroudedAbilities.map((ability) => ability.id),
});

const isZulGamuxBrokerRestorationEvent = createIsSpecificEvent<CastEvent>({
  type: "cast",
  abilityGameID: zulgamuxBrokerRestorationID,
});

const isInfiltratorBrokerRestorationEvent = createIsSpecificEvent<CastEvent>({
  type: "cast",
  abilityGameID: defaultBrokerRestorationID,
});

export const shroudedBuffs = [
  {
    id: 373_108,
    name: "Bounty: Critical Strike",
    icon: "ability_criticalstrike",
  },
  {
    id: 373_113,
    name: "Bounty: Haste",
    icon: "ability_mage_netherwindpresence",
  },
  {
    id: 373_121,
    name: "Bounty: Versatility",
    icon: "spell_arcane_arcanetactics",
  },
  {
    id: 373_116,
    name: "Bounty: Mastery",
    icon: "ability_rogue_sinistercalling",
  },
];

const isShroudedApplyBuffStackEvent =
  createIsSpecificEvent<ApplyBuffStackEvent>({
    type: "applybuffstack",
    abilityGameID: shroudedBuffs.map((buff) => buff.id),
  });

const isShroudedApplyDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: shroudedAbilities.map((buff) => buff.id),
});

export const filterExpression = [
  `type = "cast" and ability.id in (${defaultBrokerRestorationID}, ${zulgamuxBrokerRestorationID})`,
  `type = "damage" and ability.id in (${shroudedAbilities
    .map((ability) => ability.id)
    .join(", ")})`,

  `type = "applybuffstack" and ability.id in (${shroudedBuffs
    .map((buff) => buff.id)
    .join(", ")})`,

  `type = "applydebuff" and ability.id in (${shroudedAbilities
    .map((ability) => ability.id)
    .join(", ")})`,
];

export const getShroudedEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | CastEvent | ApplyBuffStackEvent | ApplyDebuffEvent)[] => {
  if (!affixSet.has(Affixes.Shrouded)) {
    return [];
  }

  return [
    ...allEvents.filter(isShroudedEvent),
    ...allEvents.filter(isZulGamuxBrokerRestorationEvent),
    ...allEvents.filter(isInfiltratorBrokerRestorationEvent),
    ...allEvents.filter(isShroudedApplyBuffStackEvent),
    ...allEvents.filter(isShroudedApplyDebuffEvent),
  ];
};
