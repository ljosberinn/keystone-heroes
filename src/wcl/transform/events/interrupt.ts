import { EventType } from "@prisma/client";

import { QUAKING } from "../../queries/events/affixes/quaking";
import type { InterruptEvent } from "../../queries/events/types";
import type { Processor } from "../utils";

export const interruptProcessor: Processor<InterruptEvent> = (
  event,
  { sourcePlayerID, targetPlayerID, targetNPCID }
) => {
  const {
    timestamp,
    extraAbilityGameID: interruptedAbilityID,
    abilityGameID,
  } = event;

  if (sourcePlayerID && targetPlayerID && abilityGameID === QUAKING) {
    return {
      timestamp,
      eventType: EventType.Interrupt,
      interruptedAbilityID,
      sourcePlayerID,
      targetPlayerID,
      abilityID: QUAKING,
    };
  }

  if (sourcePlayerID) {
    return {
      timestamp,
      eventType: EventType.Interrupt,
      interruptedAbilityID,
      sourcePlayerID,
      targetNPCID: targetNPCID ? targetNPCID : null,
      abilityID: abilityGameID,
    };
  }

  return null;
};
