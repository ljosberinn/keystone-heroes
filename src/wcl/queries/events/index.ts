import { Affixes } from "@prisma/client";

import type { DungeonIDs } from "../../../db/data/dungeons";
import type {
  AllTrackedEventTypes,
  BeginCastEvent,
  DamageEvent,
  DeathEvent,
} from "../events/types";
import { filterAffixEvents, getAffixExpression } from "./affixes";
import { findExplosiveTargetID } from "./affixes/explosive";
import {
  cheatDeathFilterExpression,
  filterCheatDeathEvents,
} from "./cheathDeath";
import {
  generalCovenantExpression,
  filterCovenantCastEvents,
} from "./covenant";
import {
  deathFilterExpression,
  filterEnemyDeathEvents,
  filterPlayerDeathEvents,
} from "./death";
import { getDungeonExpression, filterDungeonEvents } from "./dungeons";
import {
  filterPlayerInterruptEvents,
  interruptFilterExpression,
} from "./interrupts";
import {
  engineeringBattleRezExpression,
  invisibilityFilterExpression,
  filterProfessionEvents,
} from "./professions";
import {
  filterRemarkableSpellEvents,
  remarkableSpellFilterExpression,
} from "./spells";
import { filterTrinkets, trinketsFilterExpression } from "./trinkets";
import { recursiveGetEvents } from "./utils";

export type EventParams = {
  reportID: string;
  startTime: number;
  endTime: number;
  dungeonID: DungeonIDs;
  fightID: number;
  affixes: Affixes[];
};

const generateFilterExpression = ({
  dungeonID,
  affixes,
}: Pick<EventParams, "dungeonID" | "affixes">) => {
  return [
    deathFilterExpression,
    remarkableSpellFilterExpression,
    engineeringBattleRezExpression,
    invisibilityFilterExpression,
    interruptFilterExpression,
    generalCovenantExpression,
    cheatDeathFilterExpression,
    trinketsFilterExpression,
    ...getDungeonExpression(dungeonID),
    ...getAffixExpression(affixes),
  ]
    .flat()
    .map((part) => `(${part})`)
    .join(" or ");
};

export const getEvents = async (
  params: EventParams,
  playerMetaInformation: { actorID: number; class: number }[]
): Promise<{
  allEvents: AllTrackedEventTypes[];
  playerDeathEvents: DeathEvent[];
  enemyDeathEvents: (DeathEvent | BeginCastEvent | DamageEvent)[];
  explosiveTargetID: number | null;
}> => {
  const filterExpression = generateFilterExpression({
    dungeonID: params.dungeonID,
    affixes: params.affixes,
  });

  const allEvents = await recursiveGetEvents<AllTrackedEventTypes>({
    startTime: params.startTime,
    endTime: params.endTime,
    reportID: params.reportID,
    filterExpression,
  });

  const actorIDSet = new Set(
    playerMetaInformation.map((dataset) => dataset.actorID)
  );

  const dungeonEvents = filterDungeonEvents(
    allEvents,
    params.dungeonID,
    actorIDSet
  );
  const { explosiveTargetID, events: affixEvents } = filterAffixEvents(
    allEvents,
    params.affixes
  );
  const professionEvents = filterProfessionEvents(allEvents);
  const remarkableSpellEvents = filterRemarkableSpellEvents(allEvents);
  const playerDeathEvents = filterPlayerDeathEvents(
    allEvents,
    actorIDSet
    // remarkableSpellEvents
  );
  const interruptEvents = filterPlayerInterruptEvents(allEvents);
  const enemyDeathEvents = filterEnemyDeathEvents(
    allEvents,
    actorIDSet,
    params.affixes.includes(Affixes.Explosive)
      ? findExplosiveTargetID(allEvents)
      : null
  );
  const sharedCovenantAbilitiesEvents = filterCovenantCastEvents(allEvents);
  const cheatDeathEvents = filterCheatDeathEvents(allEvents);
  const trinketEvents = filterTrinkets(allEvents);

  return {
    allEvents: [
      ...professionEvents,
      ...dungeonEvents,
      ...affixEvents,
      ...remarkableSpellEvents,
      ...playerDeathEvents,
      ...interruptEvents,
      ...sharedCovenantAbilitiesEvents,
      ...cheatDeathEvents,
      ...trinketEvents,
    ].sort((a, b) => a.timestamp - b.timestamp),
    playerDeathEvents,
    enemyDeathEvents,
    explosiveTargetID,
  };
};
