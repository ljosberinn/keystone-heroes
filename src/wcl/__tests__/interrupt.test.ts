import { QUAKING } from "../queries/events/affixes/quaking";
import type { InterruptEvent } from "../queries/events/types";
import { interruptProcessor } from "../transform/events/interrupt";
import allEvents from "./fixtures/allEvents.json";

describe("interrupt", () => {
  const params = {
    sourcePlayerID: 1,
    targetPlayerID: 1,
    sourceNPCID: null,
    targetNPCID: null,
  };

  const interruptEvents = allEvents.filter(
    (event): event is InterruptEvent => event.type === "interrupt"
  );

  test("tracks interrupts by quaking", () => {
    const result = interruptEvents
      .filter((event) => event.abilityGameID === QUAKING)
      .map((event) => interruptProcessor(event, params));

    expect(result).toMatchSnapshot();
  });

  test("ignores unknown interrupt events", () => {
    expect(
      interruptProcessor(
        {
          type: "interrupt",
          timestamp: 1,
          abilityGameID: 1,
          extraAbilityGameID: 2,
          sourceID: 1,
          targetID: 1,
        },
        { ...params, sourcePlayerID: null, targetPlayerID: null }
      )
    ).toBeNull();
  });
});
