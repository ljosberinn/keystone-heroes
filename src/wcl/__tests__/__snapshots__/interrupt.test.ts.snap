// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interrupt tracks interrupts by quaking 1`] = `
[
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 212084,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 11468002,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 5782,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 11547988,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 320674,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 11607974,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 212084,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 11648002,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 323764,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 11747990,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 190984,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 11767998,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 194153,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 11907980,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 51505,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 12007992,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 51505,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 12027977,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 116858,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 12167977,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 48018,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 12447979,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 190984,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 12527973,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 8004,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 12567974,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 51505,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 12867983,
  },
  {
    "abilityID": 240448,
    "eventType": "Interrupt",
    "interruptedAbilityID": 212084,
    "sourcePlayerID": 1,
    "targetPlayerID": 1,
    "timestamp": 12987977,
  },
]
`;
